src/main/java/org/opensearch/knn/index/codec/derivedsource/DerivedSourceLuceneHelper.java

```diff
@@ -0,0 +1,238 @@
+     */
+    public List<Integer> termMatchesInRange(int startDocId, int endDocId, String termField, String term) throws IOException {
+        if (endDocId - startDocId < 0) {
+            return Collections.emptyList();
+        }
+
+        // First, we need to get the current PostingsEnum for the key as term field and term
+        PostingsEnum postingsEnum = getPostingsEnum(termField, term);
+
+        // Next, get all the docs that match this parent path. If none were found, return an empty list
+        if (postingsEnum == null) {
+            return Collections.emptyList();
+        }
+        List<Integer> matchingDocs = new ArrayList<>();
+        postingsEnum.advance(startDocId);
+        while (postingsEnum.docID() != NO_MORE_DOCS && postingsEnum.docID() <= endDocId) {
+            if (postingsEnum.freq() > 0) {
+                matchingDocs.add(postingsEnum.docID());
+            }
+            postingsEnum.nextDoc();
+        }
+
+        return matchingDocs;
+    }
+
+    /**
+     * Check if the docId is a parent for the given field. To do this, it checks if any of the documents in the range
+     * contain the parent field in the _nested_path
+     *
+     * @param offset First doc to check (inclusive)
+     * @param parentDocId document to be checked if its a parent
+     * @param parentFieldName path to parent field
+     * @return true if the docId is a parent, false otherwise
+     */
+    public boolean isNestedParent(int offset, int parentDocId, String parentFieldName) throws IOException {
+        if (parentFieldName == null) {
+            return false;
+        }
+
+        if (parentDocId <= 0) {
+            return false;
+        }
+
+        return termMatchesInRange(offset, parentDocId - 1, "_nested_path", parentFieldName).isEmpty() == false;
+    }
+
+    /**
+     * Check if the field exists for the given document.
+     *
+     * @param fieldToMatch field to check
+     * @param docId document to check
+     * @return true if the field exists for the given document, false otherwise
+     * @throws IOException if there is an issue reading from the segment
+     */
+    public boolean fieldExists(String fieldToMatch, int docId) throws IOException {
+        int firstDocId = getFirstDocWhereFieldExists(fieldToMatch, docId);
+        return firstDocId == docId;
+    }
+
+    /**
+     * Get the lowest docId for a field that is greater than (or equal to) the offset. This method is implemented in a
+     * very similar way as checking if a field exists.
+     *
+     * @param fieldToMatch field to find the lowest docId for
+     * @param offset       offset to start searching from (inclusive)
+     * @return lowest docId for the field that is greater than the offset. Returns {@link DocIdSetIterator#NO_MORE_DOCS} if doc cannot be found
+     * @throws IOException if there is an issue reading from the formats
+     */
+    public int getFirstDocWhereFieldExists(String fieldToMatch, int offset) throws IOException {
```

Comment:
Are we going to want to use this multiple times in a single request? Would it make sense to return the DocIdSetIterator instead?

Reply:
We will make this call multiple times, but for the field will not be consistent. But we could reuse if we map it to a particular field.